### ISSUE TITLE
Possible literal type inference bug

### ISSUE THREAD

**Issue by ethanresnick** [Microsoft/TypeScript#11151]
I know the rules for inference with literal types are complicated and shifting of late, so apologies if this is working as intended. The behavior's definitely not intuitive to me, though.

**TypeScript Version:**  2.1.0-dev.20160926

**Code**

```ts
type ValidOptions = "a" | "b" | "c" | undefined;

const a = <ValidOptions[]>["a"]; // assignment allowed, as expected
const d = <ValidOptions[]>["d"]; // assignment fails, as expected
const b = <ValidOptions[]>["a", "d"]; // assignment allowed. what??
```

**Expected behavior:**
`const b = <ValidOptions[]>["a", "d"];` would be a type error.

**Actual behavior:** 
`const b = <ValidOptions[]>["a", "d"];` doesn't produce an error.

**Comment by RyanCavanaugh** [comment]
You used a type assertion, which is going to allow some unsafe things to happen (otherwise there would be no point for type assertions to exist).

If you didn't use a type assertion, and instead did a normal assignability check, you'd get an error as expected:
```ts
type ValidOptions = "a" | "b" | "c" | undefined;

const a: ValidOptions[] = ["a"]; // assignment allowed, as expected
const d: ValidOptions[] = ["d"]; // assignment fails, as expected
const b: ValidOptions[] = ["a", "d"]; // assignment fails
```
The reason you were allowed to type-assert `["a", "d"]" to `ValidOptions[]` is that `Array<"a">` is assignable to `Array<"a" | "d">`.

**Comment by ethanresnick** [comment]
D'oh, that makes perfect sense. It's been a long day :)
